pkg/clusteragent/admission/mutate/autoinstrumentation/auto_instrumentation.go

```diff
@@ -329,35 +359,124 @@ func (w *Webhook) getAllLatestLibraries() []libInfo {
+	// to determine which libraries to inject.
+	libInfoSourceSingleStepInstrumentation
+	// libInfoSourceSingleStepLanguageDetection is when we use the language detection
+	// annotation to determine which libs to inject.
+	libInfoSourceSingleStepLangaugeDetection
+)
+
+// injectionType produces a string to distinguish between if
+// we're using "single step" or "lib injection" for metrics and logging.
+func (s libInfoSource) injectionType() string {
+	switch s {
+	case libInfoSourceSingleStepInstrumentation, libInfoSourceSingleStepLangaugeDetection:
+		return singleStepInstrumentationInstallType
+	case libInfoSourceLibInjection:
+		return localLibraryInstrumentationInstallType
+	default:
+		return "unknown"
 	}
+}
+
+func (s libInfoSource) isSingleStep() bool {
+	return s.injectionType() == singleStepInstrumentationInstallType
+}
+
+// isFromLanguageDetection tells us whether this source comes from
+// the language detection reporting and annotation.
+func (s libInfoSource) isFromLanguageDetection() bool {
+	return s == libInfoSourceSingleStepLangaugeDetection
+}
+
+func (s libInfoSource) mutatePod(pod *corev1.Pod) error {
+	_ = mutatecommon.InjectEnv(pod, corev1.EnvVar{
+		Name:  instrumentationInstallTypeEnvVarName,
+		Value: s.injectionType(),
+	})
+	return nil
+}
+
+type extractedPodLibInfo struct {
+	// libs are the libraries we are going to attempt to inject into the given pod.
+	libs []libInfo
+	// languageDetection is set when we ran/used the language-detection annotation.
+	languageDetection *libInfoLanguageDetection
+	// source is where we got the data from, used for telemetry later.
+	source libInfoSource
+}
+
+func (e extractedPodLibInfo) withLibs(l []libInfo) extractedPodLibInfo {
+	e.libs = l
+	return e
+}
+
+func (e extractedPodLibInfo) useLanguageDetectionLibs() (extractedPodLibInfo, bool) {
+	if e.languageDetection != nil && len(e.languageDetection.libs) > 0 && e.languageDetection.injectionEnabled {
+		e.libs = e.languageDetection.libs
+		e.source = libInfoSourceSingleStepLangaugeDetection
+		return e, true
+	}
+
+	return e, false
+}
+
+func (w *Webhook) initExtractedLibInfo(pod *corev1.Pod) extractedPodLibInfo {
+	// it's possible to get here without single step being enabled, and the pod having
+	// annotations on it to opt it into pod mutation, we disambiguate those two cases.
+	var (
+		source            = libInfoSourceLibInjection
+		languageDetection *libInfoLanguageDetection
+	)
 
-	// If auto-instrumentation is enabled in the namespace and nothing has been overridden,
-	//
-	// 1. We check for pinned libraries in the config
-	// 2. We check for language detection (if enabled)
-	// 3. We fall back to "latest"
 	if w.isEnabledInNamespace(pod.Namespace) {
-		if len(w.pinnedLibraries) > 0 {
-			return w.pinnedLibraries, false
-		}
+		source = libInfoSourceSingleStepInstrumentation
+		languageDetection = w.getLibrariesLanguageDetection(pod)
```

Comment:
Few Questions:

- Why is the source set here to `libInfoSourceSingleStepInstrumentation` and not to `libInfoSourceSingleStepLangaugeDetection`
- Why are we only checking for language detection results here, why don't we also check if the user has put annotations on the pod?